xdsl/dialects/riscv_func.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import annotations
+
+from typing import Annotated
+
+from xdsl.ir import Operation, SSAValue, Dialect, Attribute, Region
+
+from xdsl.irdl import (
+    IRDLOperation,
+    OptOpAttr,
+    OptOpResult,
+    OptOperand,
+    VarOperand,
+    irdl_op_definition,
+    SingleBlockRegion,
+    OpAttr,
+)
+from xdsl.dialects.builtin import StringAttr
+from xdsl.dialects import riscv
+from xdsl.utils.exceptions import VerifyException
+
+
+def opt_str_attr(attr: str | StringAttr | None) -> StringAttr | None:
+    if attr is None:
+        return None
+    if isinstance(attr, StringAttr):
+        return attr
+    return StringAttr(attr)
+
+
+def str_attr(attr: str | StringAttr) -> StringAttr:
+    if isinstance(attr, StringAttr):
+        return attr
+    return StringAttr(attr)
+
+
+@irdl_op_definition
+class CallOp(IRDLOperation):
+    name = "riscv_func.call"
+    args: Annotated[VarOperand, riscv.RegisterType]
+    func_name: OpAttr[StringAttr]
+    result: Annotated[OptOpResult, riscv.RegisterType]
+
+    def __init__(
+        self,
+        func_name: StringAttr,
+        args: list[Operation | SSAValue],
+        has_result: bool = True,
+        comment: StringAttr | None = None,
+    ):
+        super().__init__(
+            operands=[args],
+            result_types=[riscv.RegisterType(riscv.Register()) if has_result else None],
+            attributes={
+                "func_name": func_name,
+                "comment": comment,
+            },
+        )
+
+
+@irdl_op_definition
+class FuncOp(IRDLOperation):
+    name = "riscv_func.func"
+
+    func_name: OpAttr[StringAttr]
+    func_body: SingleBlockRegion
+
+    def __init__(self, name: str, region: Region):
+        attributes: dict[str, Attribute] = {"func_name": StringAttr(name)}
+
+        super().__init__(attributes=attributes, regions=[region])
+
+    def verify_(self):
+        # Check that the returned value matches the type of the function
+        if len(self.func_body.blocks) != 1:
+            raise VerifyException("Expected FuncOp to contain one block")
+
+        block = self.func_body.blocks[0]
+
+        if not len(block.ops):
+            raise VerifyException("Expected FuncOp to not be empty")
+
+        last_op = block.last_op
+
+        if not isinstance(last_op, ReturnOp):
+            raise VerifyException("Expected last op of FuncOp to be a ReturnOp")
+
+        # Sasha: the following is copy/pasted code from the Toy dialect that checks
```

Comment:
Please remove commented code, and add two verifiers:

 - Number of arguments must be less than or equal to 8
 - Number of results must be less than or equal to two

The syscall PR has similar checks

Reply:
actually we could probably check that the return operation in this function also has the same number of operands as this function type has results

Reply:
> actually we could probably check that the return operation in this function also has the same number of operands as this function type has results

Added this as a todo